src/Server/middleware/graphqlProxyMiddleware.ts

```diff
@@ -0,0 +1,207 @@
+import { createProxyMiddleware, fixRequestBody } from "http-proxy-middleware"
+import {
+  METAPHYSICS_ENDPOINT,
+  GRAPHQL_CACHE_TTL,
+  ENABLE_GRAPHQL_CACHE,
+} from "Server/config"
+import { cache } from "Server/cacheClient"
+import { createBrotliDecompress, createGunzip } from "zlib"
+import { ArtsyRequest, ArtsyResponse } from "Server/middleware/artsyExpress"
+import { NextFunction } from "express"
+import { IncomingMessage } from "http"
+import {
+  RELAY_CACHE_CONFIG_HEADER_KEY,
+  RELAY_CACHE_PATH_HEADER_KEY,
+} from "System/Relay/middleware/cacheHeaderMiddleware"
+import { findRoutesByPath } from "System/Router/Utils/routeUtils"
+
+export const graphqlProxyMiddleware = async (
+  req: ArtsyRequest,
+  res: ArtsyResponse,
+  next: NextFunction
+) => {
+  const skipCache = !!req.user || shouldSkipCache(req)
+
+  if (!skipCache) {
+    const cachedResponse = await readCache(req)
+
+    if (cachedResponse) {
+      return res.json(cachedResponse)
+    }
+  }
+
+  return createProxyMiddleware({
+    target: `${METAPHYSICS_ENDPOINT}/v2`,
+    changeOrigin: true,
+    on: {
+      // Fix proxy incompatability with express.js body-parser middleware
+      proxyReq: fixRequestBody,
+      proxyRes: (
+        proxyRes: IncomingMessage,
+        req: ArtsyRequest,
+        res: ArtsyResponse
+      ) => {
+        if (!skipCache) {
+          writeCache(proxyRes, req, res)
+        }
+      },
+    },
+  })(req, res, next)
+}
+
+export const readCache = async (req: ArtsyRequest) => {
+  if (ENABLE_GRAPHQL_CACHE) {
+    try {
+      const queryId = req.body?.id
+      const variables = req.body?.variables
+
+      const cacheKey = getCacheKey({
+        queryId,
+        variables,
+      })
+
+      const response = await cache.get(cacheKey)
+
+      if (response) {
+        console.log(
```

Comment:
Do we need to log all of these (wonder who will look at these logs, they fill up Papertrail).

Possibly better a metric we could look at in Datadog (cache hit rate)?

Reply:
Lets keep them around for a bit as we release just for debugging (papertrail, lol) reasons, but yah we can maybe log only in dev after a bit